lazy init device mesh in fsdp

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104056
* __->__ #104447

since fsdp state is lazy init, we also need to lazy init device mesh
otherwise devicemesh allgather check would trigger some mismatch in
allgather counts in fsdp tests

wanchaol committed:
lazy init device mesh in fsdp

since fsdp state is lazy init, we also need to lazy init device mesh
otherwise devicemesh allgather check would trigger some mismatch in
allgather counts in fsdp tests

[ghstack-poisoned]

wanchaol committed:
Update on "lazy init device mesh in fsdp"


since fsdp state is lazy init, we also need to lazy init device mesh
otherwise devicemesh allgather check would trigger some mismatch in
allgather counts in fsdp tests

[ghstack-poisoned]

wconstab commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -226,9 +226,12 @@ def _init_device_mesh(
     mesh_tensor = torch.arange(get_world_size(root_state.process_group))
-    device_mesh = DeviceMesh(device_type, mesh_tensor)
+    device_mesh = DeviceMesh(device_type, mesh_tensor, _init_process_groups=False)
+    device_mesh._dim_group_infos = (_get_group_tag(root_state.process_group), mesh_tensor.tolist())
```
i'm still wondering if this part should be hidden inside DeviceMesh ctor if we pass pg to it.  But that can be in another PR

wanchaol replied:
Yeah I'll do a follow up PR to address this

wconstab reviewed (approved):
lgtm. and i verified it passes most relevant distributed tests on https://github.com/pytorch/pytorch/pull/104056 locally. have to run with CI to verify the rest..

wanchaol commented:
@pytorchbot merge

wanchaol committed:
Update on "lazy init device mesh in fsdp"


since fsdp state is lazy init, we also need to lazy init device mesh
otherwise devicemesh allgather check would trigger some mismatch in
allgather counts in fsdp tests

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -226,9 +226,15 @@ def _init_device_mesh(
         return None
     if get_backend() == "fake" or not root_state.compute_device:
         return None
+    from torch.distributed.distributed_c10d import _get_group_tag
```
Is this intentionally imported here and not at the top?

wanchaol replied:
not really, I planned to address the lazy init in a follow up PR so this import might not needed later, I can move it to top first.

wanchaol committed:
Update on "lazy init device mesh in fsdp"


since fsdp state is lazy init, we also need to lazy init device mesh
otherwise devicemesh allgather check would trigger some mismatch in
allgather counts in fsdp tests

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
